### PR TITLE
Add prediction saving script

### PR DESCRIPTION
**Pull request by jeswan**
It is highly recommended that the script is run with a GPU! Approximately 60x faster.

### CODE REVIEW

File: src/data/right_data.py [hunk 1]
```diff
@@ -11,9 +11,85 @@ def __init__(self, *args, **kwargs):
...
 
+    def pred_dataloader(self):
+        loader = DataLoader(
+            self.pred_dataset,
+            batch_size=self.batch_size,
+            shuffle=False,
+            num_workers=self.num_workers,
+        )
+        return loader
+
```
Comment (jeswan) [inline src/data/right_data.py]:
```suggestion
```

File: scripts/postproc/generate_predictions.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    model = DirectRightModel.load_from_checkpoint(checkpoint_path=args.ckpt)
+    model.to(device)
+    model.eval()
+    model.freeze()
+
+    print("model instance created")
+    print("lightning version", pl.__version__)
+
+    outputs = []
+    for idx, batch in enumerate(tqdm(dm.val_dataloader())):
```
Comment (jeswan) [inline scripts/postproc/generate_predictions.py:68]: we currently just use the val_dataloader (that has not been shuffled)

File: scripts/postproc/generate_predictions.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    print("size of prediction samples:", len(dm.val_samples))
+
+    img, target, extra_info = next(iter(dm.val_dataloader()))
+    print(img.shape)
+    print(target.shape)
+    print(f"extra info len: {len(extra_info)}")
+
+    # model
+    model = DirectRightModel(**args.__dict__)
+    model = DirectRightModel.load_from_checkpoint(checkpoint_path=args.ckpt)
```
Comment (jeswan) [inline scripts/postproc/generate_predictions.py]: Could use some input here...the checkpoint won't load any DirectRightModel class variables correct? Wondering if there is an easy way to save them in the checkpoint. We could also pass in the parameters when calling the script.

File: scripts/postproc/generate_predictions.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        # concat left view and right view
+        left_and_right = torch.cat((img, preds), dim=3).permute(0,2,3,1)
+        left_and_right = torch.clamp(left_and_right, 0, 1) * 255
+        outputs.append(left_and_right)
+
+        # output in chunks to avoid memory errors
+        if (idx + 1) % 20 == 0:
+            outputs = torch.cat(outputs)
+            torchvision.io.write_video(filename=os.path.join(args.output_dir, f"{idx}.avi"), video_array=outputs, fps=30)
+            outputs = []
```
Comment (jeswan) [inline scripts/postproc/generate_predictions.py:82]:
```suggestion
            outputs = []
        outputs = torch.cat(outputs)
        torchvision.io.write_video(filename=os.path.join(args.output_dir, f"{idx}.avi"), video_array=outputs, fps=30)
```

File: scripts/postproc/generate_predictions.py (hunk 3 above)
Comment (annikabrundyn) [reply scripts/postproc/generate_predictions.py]: the checkpoint does load the parameters :)

File: src/data/right_data.py [hunk 5]
```diff
@@ -17,3 +17,70 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
         self.dataset = RightDaVinciDataSet
+
```
Comment (annikabrundyn) [inline src/data/right_data.py:20]: why did you need to add this?

Comment (annikabrundyn) [reply src/data/right_data.py:20]: it looks pretty much the same as setup() no?

File: scripts/postproc/generate_predictions.py (hunk 3 above)
Comment (jeswan) [reply scripts/postproc/generate_predictions.py]: excellent!

File: scripts/postproc/generate_predictions.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    # sanity check
+    print("size of prediction samples:", len(dm.val_samples))
+
+    img, target, extra_info = next(iter(dm.val_dataloader()))
+    print(img.shape)
+    print(target.shape)
+    print(f"extra info len: {len(extra_info)}")
+
+    # model
+    model = DirectRightModel(**args.__dict__)
```
Comment (jeswan) [inline scripts/postproc/generate_predictions.py]:
```suggestion
```

File: src/data/right_data.py (hunk 5 above)
Comment (annikabrundyn) [reply src/data/right_data.py:20]: might be worth just writing right data not as subclass?

File: scripts/postproc/generate_predictions.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    dm = RightDaVinciDataModule(
+        args.data_dir,
+        frames_per_sample=args.frames_per_sample,
+        frames_to_drop=args.frames_to_drop,
+        include_right_view=args.include_right_view,
+        stack_horizontal=args.stack_horizontal,
+        is_color_input=args.is_color_input,
+        extra_info=True,
+        batch_size=args.batch_size,
+        num_workers=args.num_workers,
```
Comment (jeswan) [inline scripts/postproc/generate_predictions.py:45]: I think my comment below may apply here then. Seems like the best way forward would be to load the model and use the class variables to set these.

File: src/data/right_data.py (hunk 5 above)
Comment (jeswan) [reply src/data/right_data.py:20]: it's `setup()` without the shuffle in the middle
removed:
```
        # shuffle all 41 sets of 1000 frames
        self.all_sets = shuffle(all_sets, random_state=42)
```
and
```
        self.train_samples = shuffle(self.train_samples, random_state=42)
        self.val_samples = shuffle(self.val_samples, random_state=42)
        self.test_samples = shuffle(self.test_samples, random_state=42)
```
